### PR TITLE
feat(bazel): @angular/bazel compat with bazel 0.27 and rules_nodejs 0.32.2

### PR DESCRIPTION
**Pull request by gregmagolan**
* Updates to Bazel 0.27.0 and makes compatibility fixes.
* Updates to rules_nodejs 0.32.2 and makes compatibility fixes for breaking changes since 0.31.x
* Turns on symlinked node_modules finally as 0.32.2 fixes rules_nodejs#802
* Fixes worker mode issue with ts_library (yeay) -- ts_library workers should be turned back on
* Improves how ts_library consumes @types npm packages
***See 0.32.2 release notes for more info: https://github.com/bazelbuild/rules_nodejs/releases/tag/0.32.2***

## rules_nodejs 0.32.0 compat fixes

@npm//node_modules/foobar:foobar.js labels changed to @npm//:node_modules/foobar/foobar.js with fix for bazelbuild/rules_nodejs#802

## Bazel 0.27.0 compat fixes

ctx.new_file is deprecated and ctx.actions.declare_file should be used in Bazel 0.27.0. Error seen here is:
```
        File "ng_module.bzl", line 272, in _expected_outs
                ctx.new_file(ctx.genfiles_dir, (ctx.label.name ..."))
Use ctx.actions.declare_file instead of ctx.new_file.
Use --incompatible_new_actions_api=false to temporarily disable this check.
```
This can be worked around with incompatible_new_actions_api flag but may as well fix it proper so downstream doesn't require this flag due to this code.

Also, depset() is no longer iterable by default without a flag. This required fixing in a few spots in @angular/bazel.

## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?

- [ ] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [x] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] angular.io application / infrastructure changes
- [ ] Other... Please describe:

## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No

### CODE REVIEW

**Comment by gregmagolan** [comment]
After updating CircleCI to node 10.14 seeing some strange RBE errors in CI:

```
#!/bin/bash -eo pipefail
openssl aes-256-cbc -d -in .circleci/gcp_token -k "$CI_REPO_NAME" -out /home/circleci/.gcp_credentials
echo "export GOOGLE_APPLICATION_CREDENTIALS=/home/circleci/.gcp_credentials" >> $BASH_ENV
sudo bash -c "echo 'build --config=remote' >> /etc/bazel.bazelrc"
bad decrypt
140224540337408:error:06065064:digital envelope routines:EVP_DecryptFinal_ex:bad decrypt:../crypto/evp/evp_enc.c:535:
Exited with code 1
```

**Comment by filipesilva** [comment]
@gregmagolan you'll need to add a flag to `openssl` to make it compatible, see https://github.com/angular/angular-cli/pull/14708.

**Review by alexeagle — COMMENTED** [review]
What if we do the breaking change fix separately from updating Bazel?

File: .circleci/config.yml [hunk 1]
```diff
@@ -15,8 +15,10 @@
 #             `CI_CHROMEDRIVER_VERSION_ARG` env var (in `.circleci/env.sh`) points to a ChromeDriver
 #             version that is compatible with the Chrome version in the image.
 # **NOTE 2**: If you change the version of the docker images, also change the `cache_key` suffix.
-var_1: &default_docker_image circleci/node:10.12
-var_2: &browsers_docker_image circleci/node:10.12-browsers
+var_1: &default_docker_image circleci/node:10.14
+var_2: &browsers_docker_image circleci/node:10.14-browsers
+# node:10.14-browsers comes with Google Chrome 71.0.3578.98 which is compatible with ChromeDriver 2.45
```
Comment (devversion) [inline .circleci/config.yml]: Nit: should the comment be above the actual anchor to which the comment refers to?

**Review by devversion — APPROVED** [review]
LGTM in general. Needs cleanup for the openssl failures.

File: .circleci/config.yml (hunk 1 above)
Comment (gregmagolan) [reply .circleci/config.yml]: 👍

**Comment by gregmagolan** [comment]
Should wait until https://github.com/angular/angular/pull/31088 is landed and then rebase on that as it will conflict

**Comment by gregmagolan** [comment]
Also fixes 0.27.0 compat break:

```
Traceback (most recent call last):
	File "/home/circleci/ng/packages/compiler-cli/test/ngtsc/fake_core/BUILD.bazel", line 13
		ng_package(name = 'npm_package')
	File "/home/circleci/ng/packages/bazel/src/ng_package/ng_package.bzl", line 301, in _ng_package_impl
		fesm2015.append(_rollup(ctx, "fesm2015", esm2015...))
	File "/home/circleci/ng/packages/bazel/src/ng_package/ng_package.bzl", line 301, in fesm2015.append
		_rollup(ctx, "fesm2015", esm2015_config, es201..., <2 more arguments>)
	File "/home/circleci/ng/packages/bazel/src/ng_package/ng_package.bzl", line 143, in _rollup
		ctx.actions.run(progress_message = ("ng_package:...)), <5 more arguments>)
Found tool(s) 'bazel-out/host/bin/external/build_bazel_rules_nodejs/internal/rollup/rollup' in inputs. A tool is an input with executable=True set. All tools should be passed using the 'tools' argument instead of 'inputs' in order to make their runfiles available to the action. This safety check will not be performed once the action is modified to take a 'tools' argument. To temporarily disable this check, set --incompatible_no_support_tools_in_action_inputs=false.
```

**Review by IgorMinar — COMMENTED** [review]
I'd mark this as feat rather than fix.

The change looks good except for a few things:
- if we are going to roll this out and require users to rm node_modules how will we notify them of this requirement? What will happen if they don't do it?
- what's up with service-worker.d.ts being excluded as part of this PR?
- do we now expect users to manage and update their global Bazel version and expect them to remove it from package.json? I see that you didn't remove it, which makes me unclear about the interactions between global Bazel and Bazel installed @bazel/bazel

**Comment by gregmagolan** [comment]
Thanks for taking a look @IgorMinar 

> I'd mark this as feat rather than fix.
Sounds good. Scope for PR increased now that we released rules_nodejs 0.32.0 today which has important fixes.
 
> The change looks good except for a few things:
> * if we are going to roll this out and require users to rm node_modules how will we notify them of this requirement? What will happen if they don't do it?
I need to think about all of the possibilities here but no rush on landing this as we likely want to wait until @bazel/bazel 0.27.1 is pushed to npm which will have an implicit dep on @bazel/hide-bazel-files which will be required.

> * what's up with service-worker.d.ts being excluded as part of this PR?
WIP. The angular/angular repo exposed an issue of type duplicates between `typescript/lib/webworker.d.ts` and `service-worker.d.ts` introduced in rules_nodejs 0.32.0. We'll need a rules_nodejs patch release 0.32.1 to fix it and add test coverage to rules_nodejs. Still working on the proper shape of that.

> * do we now expect users to manage and update their global Bazel version and expect them to remove it from package.json? I see that you didn't remove it, which makes me unclear about the interactions between global Bazel and Bazel installed @bazel/bazel
User's still have the choice to user either or. The happy path is to use @bazel/bazel for front-end users and this is what angular-cli sets up in the bazel opt-in.

**Comment by gregmagolan** [comment]
Also disables `must-use-promises` tsetse check in ts_library tsetse for tests as, with 0.32.0, tsetse now falsely asserting on some lines in a few tests such as `packages/core/schematics/test/injectable_pipe_migration_spec.ts`.

```
await runMigration();
expect(tree.readContent('/index.ts'))
```

tsetse asserts that "await is required on promise" on the 2nd line when there is no promise there.

@alexeagle Could adding of .d.ts files to the tsconfig explicitly in 0.32.0 have caused this?

**Comment by IgorMinar** [comment]
The thing that I don't understand is - will we still be asking users to
install bazel twice? and how will that work with the requirement to `rf
node_modules`.

This is the workflow that puzzles me:
- yarn install
- rm -rf node_modules
- yarn bazel build packages/...

this should error because bazel was deleted in the `rm -rf` step. What am I
missing here?

File: packages/tsconfig-test.json [hunk 2]
```diff
@@ -5,6 +5,10 @@
   "extends": "./tsconfig-build.json",
   "compilerOptions": {
     "types": ["node", "jasmine"],
+    "plugins": [{
+        "name": "@bazel/tsetse",
+        "disabledRules": ["must-use-promises"],
```
Comment (alexeagle) [inline packages/tsconfig-test.json]: why not just fix the code that doesn't use the promise?

**Comment by gregmagolan** [comment]
> why not just fix the code that doesn't use the promise?

@alexeagle I looked at it and the code is not using a promise. It looks like a tsetse bug on this assertion. Possibly introduced now that `.d.ts` files are added to the tsconfig?

**Comment by gregmagolan** [comment]
@IgorMinar Yes. It's a bit confusing as a result of fixing https://github.com/bazelbuild/rules_nodejs/issues/802.

The good news is that its fixed we can switch angular and angular-cli bazel-opt in to use symlinked node_modules now. Working in a 0.32.1 right now for this as the angular/angular repo exposed so issues with ts_library and tsetse that did not come up in rules_nodejs.

The bad news is that for users updating from rules_node 0.31.x should delete their `node_modules` folder as one-time step when upgrading to 0.32.x. `node_modules` had `yarn_install`/`npm_install` generated `BUILD.bazel` in it (when symlinking only so this won't apply to dev-infra angular/angular devs) and these will cause build errors after the upgrade.

Steps for users would be

```
rm -rf node_modules // one time-only
yarn install
yarn bazel build //...
```

or if you have bazel installed locally just,

```
rm -rf node_modules // one time-only
bazel build //...
```

angular-devs won't need to do this since I haven't turned on node_modules symlinked in this repo yet.

@kyliau Is is possible in bazel schematics to delete the user's node_modules folder if they were on 0.31.x and upgrading to 0.32.x? If so, should we do this for angular-cli bazel opt-in users that us `ng upgrade`?

**Comment by gregmagolan** [comment]
> will we still be asking users to install bazel twice?

@IgorMinar Realized that I didn't answer that question. The answer is no, only once is required now that node_modules are symlinked :)

**Comment by gregmagolan** [comment]
Required fixes to ts_library made here https://github.com/bazelbuild/rules_nodejs/pull/875. Waiting on that to land and a 0.32.1 release.

**Review by alexeagle — APPROVED** [review]

**Comment by kara** [comment]
@gregmagolan This doesn't apply cleanly to patch. Can you create a separate PR?

**Comment by gregmagolan** [comment]
> @gregmagolan This doesn't apply cleanly to patch. Can you create a separate PR? Merging this one to master.

Sure. I just rebased as well to fix a yarn.lock conflict so CI running again.

**Comment by kara** [comment]
@gregmagolan Re-opened this PR so we can work on it after the rollback. Will need presubmit.

**Comment by kyliau** [comment]
g3 presubmit: 	http://test/OCL:255445653:BASE:255445684:1561660930172:b1953726

**Comment by gregmagolan** [comment]
Replaced by #31325

**Pull request closed without merging**